File: internal/lib/metadata_model/database_field_value.go
package metadatamodel

import (
	"errors"
	"reflect"

	intlibjson "github.com/icipe-official/Data-Abstraction-Platform/internal/lib/json"
)

func DatabaseGetColumnFieldValue(metadatamodel any, columnFieldName string, tableCollectionUID string, valueToGetFrom any, skipIfFGDisabled bool, skipIfDataExtraction bool) (any, error) {
	if len(columnFieldName) == 0 {
		return nil, FunctionNameAndError(DatabaseGetColumnFieldValue, errors.New("columnFieldName is empty"))
	}

	databaseColumnFields, err := DatabaseGetColumnFields(metadatamodel, tableCollectionUID, skipIfFGDisabled, skipIfDataExtraction)
	if err != nil {
		return nil, FunctionNameAndError(DatabaseGetColumnFieldValue, err)
	}

	if columnField, ok := databaseColumnFields.Fields[columnFieldName]; ok {
		if columnFieldMap, ok := columnField.(map[string]any); ok {
			pathToColumnFieldValue, err := GetValueAsString(columnFieldMap[FIELD_2D_POSITION_PROP_FIELD_GROUP_KEY])
			if err != nil {
				return nil, FunctionNameAndError(DatabaseGetColumnFieldValue, err)
			}
			pathToColumnFieldValue = GetPathToValue(pathToColumnFieldValue, true)

			return intlibjson.GetValueInObject(valueToGetFrom, pathToColumnFieldValue), nil
		} else {
			return nil, argumentsError(DatabaseGetColumnFieldValue, "columnFieldMap", "map[string]any", columnField)
		}
	} else {
		return nil, FunctionNameAndError(DatabaseGetColumnFieldValue, errors.New("columnField is empty"))
	}
}

func DatabaseDeleteColumnFieldValue(metadatamodel any, columnFieldName string, tableCollectionUID string, valueToDeleteIn any, skipIfFGDisabled bool, skipIfDataExtraction bool) (any, error) {
	if len(columnFieldName) == 0 {
		return nil, FunctionNameAndError(DatabaseDeleteColumnFieldValue, errors.New("columnField is empty"))
	}

	databaseColumnFields, err := DatabaseGetColumnFields(metadatamodel, tableCollectionUID, skipIfFGDisabled, skipIfDataExtraction)
	if err != nil {
		return nil, FunctionNameAndError(DatabaseDeleteColumnFieldValue, err)
	}

	if columnField, ok := databaseColumnFields.Fields[columnFieldName]; ok {
		if columnFieldMap, ok := columnField.(map[string]any); ok {
			pathToColumnFieldValue, err := GetValueAsString(columnFieldMap[FIELD_2D_POSITION_PROP_FIELD_GROUP_KEY])
			if err != nil {
				return nil, FunctionNameAndError(DatabaseDeleteColumnFieldValue, err)
			}
			pathToColumnFieldValue = GetPathToValue(pathToColumnFieldValue, true)

			return intlibjson.DeleteValueInObject(valueToDeleteIn, pathToColumnFieldValue), nil
		} else {
			return nil, argumentsError(DatabaseDeleteColumnFieldValue, "columnFieldMap", "map[string]any", columnField)
		}
	} else {
		return nil, FunctionNameAndError(DatabaseDeleteColumnFieldValue, errors.New("columnField is empty"))
	}
}

func DatabaseSetColumnFieldValue(metadatamodel any, columnFieldName string, tableCollectionUID string, valueToGetIn any, value any, skipIfFGDisabled bool, skipIfDataExtraction bool) (any, error) {
	if len(columnFieldName) == 0 {
		return nil, FunctionNameAndError(DatabaseSetColumnFieldValue, errors.New("columnField is empty"))
	}

	databaseColumnFields, err := DatabaseGetColumnFields(metadatamodel, tableCollectionUID, skipIfFGDisabled, skipIfDataExtraction)
	if err != nil {
		return nil, FunctionNameAndError(DatabaseSetColumnFieldValue, err)
	}

	if columnField, ok := databaseColumnFields.Fields[columnFieldName]; ok {
		if columnFieldMap, ok := columnField.(map[string]any); ok {
			pathToColumnFieldValue, err := GetValueAsString(columnFieldMap[FIELD_2D_POSITION_PROP_FIELD_GROUP_KEY])
			if err != nil {
				return nil, FunctionNameAndError(DatabaseSetColumnFieldValue, err)
			}
			pathToColumnFieldValue = GetPathToValue(pathToColumnFieldValue, true)

			if reflect.TypeOf(value).Kind() == reflect.Slice || reflect.TypeOf(value).Kind() == reflect.Array {
				return intlibjson.SetValueInObject(valueToGetIn, pathToColumnFieldValue, value)
			}

			return intlibjson.SetValueInObject(valueToGetIn, pathToColumnFieldValue, []any{value})
		} else {
			return nil, argumentsError(DatabaseSetColumnFieldValue, "columnFieldMap", "map[string]any", columnField)
		}
	} else {
		return nil, FunctionNameAndError(DatabaseSetColumnFieldValue, errors.New("columnField is empty"))
	}
}

File: web/src/lib/metadata_model/database_field_value.ts
import Json from '../json'
import MetadataModel from '.'


/**
 * Sets the value of a field with database properties
 *
 * @throws {MetadataModel.Error} if arguments are not vald or column is not found.
 *
 * @param metadatamodel Expected to be presented as if converted from JSON.
 * @param columnFieldName database column/field to match against retrieved Database Column Fields when {@linkcode_MetadataModel.DatabaseGetColumnFields} is called.
 * @param tableCollectionName Extract only fields whose {@linkcode FgProperties.DATABASE_TABLE_COLLECTION_NAME} match this value.
 * @param valueToSetIn
 * @param value
 * @param skipIfFGDisabled Do not include field if property {@linkcode FgProperties.FIELD_GROUP_VIEW_DISABLE}($FG_VIEW_DISABLE) is true. Default false.
 * @param skipIfDataExtraction Do not include field if property {@linkcode FgProperties.FIELD_GROUP_SKIP_DATA_EXTRACTION}($FG_SKIP_DATA_EXTRACTION) is true. Default false.
 * @returns Updated {@linkcode valueToSetIn}
 */
export function DatabaseSetColumnFieldValue(metadatamodel: any, columnFieldName: string, tableCollectionUID: string, valueToSetIn: any, value: any, skipIfFGDisabled: boolean = false, skipIfDataExtraction: boolean = false) {
	if (!MetadataModel.IsGroupFieldsValid(metadatamodel)) {
		throw [DatabaseSetColumnFieldValue.name, 'argument metadatamodel is not an object']
	}

	if (typeof columnFieldName !== 'string') {
		throw [DatabaseSetColumnFieldValue.name, 'argument columnFieldName is not a string']
	}

	if (typeof tableCollectionUID !== 'string') {
		throw [DatabaseSetColumnFieldValue.name, 'argument tableCollectionUID is not a string']
	}

	let databseColumnFields: MetadataModel.IDatabaseColumnFields
	try {
		databseColumnFields = MetadataModel.DatabaseGetColumnFields(metadatamodel, tableCollectionUID, skipIfFGDisabled, skipIfDataExtraction)
	} catch (e) {
		throw e
	}

	let columnField = databseColumnFields.fields[columnFieldName]
	if (typeof columnField !== 'object') {
		throw [DatabaseSetColumnFieldValue.name, 'column ${columnField} not found among list of databseColumnFields.fields']
	}

	const pathToColumnFieldValue = (columnField[MetadataModel.FgProperties.FIELD_GROUP_KEY] as string)
		.replace(/\.\$GROUP_FIELDS\[\*\]/, '')
		.replace(new RegExp(MetadataModel.GROUP_FIELDS_REGEX_SEARCH, 'g'), '')
		.replace(new RegExp(MetadataModel.ARRAY_PATH_REGEX_SEARCH, 'g'), '[0]')

	try {
		return Json.SetValueInObject(valueToSetIn, pathToColumnFieldValue, Array.isArray(value) ? value : [value])
	} catch (e) {
		throw e
	}
}

export function DatabaseGetColumnFieldValue(metadatamodel: any, columnFieldName: string, tableCollectionUID: string, valueToGetFrom: any, skipIfFGDisabled: boolean = false, skipIfDataExtraction: boolean = false) {
    if (!MetadataModel.IsGroupFieldsValid(metadatamodel)) {
        throw 'argument metadatamodel is not an object.'
    }

    if (typeof columnFieldName !== 'string') {
        throw 'argument columnFieldName is not a string.'
    }

    if (typeof tableCollectionUID !== 'string') {
        throw 'argument tableCollectionUID is not a string.'
    }

    let databseColumnFields: MetadataModel.IDatabaseColumnFields
    try {
        databseColumnFields = MetadataModel.DatabaseGetColumnFields(metadatamodel,tableCollectionUID, skipIfFGDisabled, skipIfDataExtraction)
    } catch (e) {
        throw e
    }

    let columnField = databseColumnFields.fields[columnFieldName]
    if (typeof columnField !== 'object') {
        throw `column ${columnField} not found among list of databseColumnFields.fields`
    }

    const pathToColumnFieldValue = (columnField[MetadataModel.FgProperties.FIELD_GROUP_KEY] as string)
        .replace(/\.\$GROUP_FIELDS\[\*\]/, '')
        .replace(new RegExp(MetadataModel.GROUP_FIELDS_REGEX_SEARCH, 'g'), '')
        .replace(new RegExp(MetadataModel.ARRAY_PATH_REGEX_SEARCH, 'g'), '[0]')

    try {
        return Json.GetValueInObject(valueToGetFrom, pathToColumnFieldValue)
    } catch (e) {
        throw e
    }
}

export function DatabaseDeleteColumnFieldValue(metadatamodel: any, columnFieldName: string, tableCollectionUID: string, valueToGetIn: any, skipIfFGDisabled: boolean = false, skipIfDataExtraction: boolean = false) {
    if (!MetadataModel.IsGroupFieldsValid(metadatamodel)) {
        throw 'argument metadatamodel is not an object.'
    }

    if (typeof columnFieldName !== 'string') {
        throw 'argument columnFieldName is not a string.'
    }

    if (typeof tableCollectionUID !== 'string') {
        throw 'argument tableCollectionUID is not a string.'
    }

    let databseColumnFields: MetadataModel.IDatabaseColumnFields
    try {
        databseColumnFields = MetadataModel.DatabaseGetColumnFields(metadatamodel, tableCollectionUID, skipIfFGDisabled, skipIfDataExtraction)
    } catch (e) {
        throw e
    }

    let columnField = databseColumnFields.fields[columnFieldName]
    if (typeof columnField !== 'object') {
        throw `column ${columnField} not found among list of databseColumnFields.fields`
    }

    const pathToColumnFieldValue = (columnField[MetadataModel.FgProperties.FIELD_GROUP_KEY] as string)
        .replace(/\.\$GROUP_FIELDS\[\*\]/, '')
        .replace(new RegExp(MetadataModel.GROUP_FIELDS_REGEX_SEARCH, 'g'), '')
        .replace(new RegExp(MetadataModel.ARRAY_PATH_REGEX_SEARCH, 'g'), '[0]')

    try {
        return Json.DeleteValueInObject(valueToGetIn, pathToColumnFieldValue)
    } catch (e) {
        throw e
    }
}
